src/python/pants/backend/go/target_type_rules.py

```diff
@@ -24,8 +46,103 @@ async def inject_go_module_dependency(request: InjectGoModuleDependency) -> Inje
+    candidate_go_external_module_targets = await Get(
+        UnexpandedTargets, AddressSpecs([MaybeEmptyDescendantAddresses("")])
+    )
+    go_external_module_targets = [
+        tgt for tgt in candidate_go_external_module_targets if isinstance(tgt, GoExternalModule)
+    ]
+    third_party_import_path_to_address = {
+        tgt[GoImportPath].value: tgt.address for tgt in go_external_module_targets
+    }
+
+    # Resolve all of the packages found.
+    first_party_import_path_to_address = {}
+    first_party_go_packages = await MultiGet(
+        Get(ResolvedGoPackage, ResolveGoPackageRequest(tgt.address)) for tgt in go_package_targets
+    )
+    for first_party_go_package in first_party_go_packages:
+        # Skip packages that are not part of this package's module.
+        # TODO: This requires that all first-party code in the monorepo be part of the same go_module. Will need
+        # figure out how multiple modules in a monorepo can interact.
+        if first_party_go_package.module_address != this_go_package.module_address:
+            continue
+
+        first_party_import_path_to_address[
+            first_party_go_package.import_path
+        ] = first_party_go_package.address
+
+    # Loop through all of the imports of this package and add dependencies on other packages and
+    # external modules.
+    inferred_dependencies = []
+    for import_path in this_go_package.imports + this_go_package.test_imports:
+        # Check whether the import path comes from the standard library.
+        if import_path in goroot_imports.import_path_mapping:
+            continue
+
+        # Infer first-party dependencies to other packages in same go_module.
+        if import_path in first_party_import_path_to_address:
+            inferred_dependencies.append(first_party_import_path_to_address[import_path])
+            continue
+
+        # Infer third-party dependencies on go_external_module targets.
+        found_module_import = False
+        for module_import_path, address in third_party_import_path_to_address.items():
+            # TODO: This check assumes that the import path for the external module is a prefix of any package
+            # from the external module and that the import path will never overlap with another external module's
+            # import path. This may be the case, but bears confirmation.
+            #
+            # TODO: Also cleanup mandatory go_module ownership, so module_import_path is not Optional[str].
+            if module_import_path and import_path.startswith(module_import_path):
+                inferred_dependencies.append(address)
+                found_module_import = True
+
+        if found_module_import:
+            continue
+
+        raise ValueError(
+            f"Unable to infer dependency for import path '{import_path}' "
+            f"in go_package at address '{this_go_package.address}'."
+        )
```

Comment:
Inference for other languages currently doesn't fail when a match cannot be found (and assumes an explicit dependency will be added instead): do we have high enough confidence in our ability to infer things that we can afford to fail here?

One reason to not fail is that erroring eagerly like this will only show you one error message: and when something is broken like this, it's likely that there is more than one error. And the compiler is in a reasonably good spot to render them.

Reply:
Will make it not fail.